### ISSUE TITLE
Allow non-SQL (un-)deployment scripts

### ISSUE THREAD

**Issue by mlawren** [theory/sqitch#1]
[I'm not sure where you are tracking issue or feature requests so for now here is as good as anywhere I guess]

It may be necessary to use scripts written in languages other than SQL (eg Perl) to perform certain steps when upgrading or downgrading a database. These scripts should also be part of the dependency graph, and of course indempotent. Database connection parameters could be passed in as environment variables or as yaml/json on stdin etc.

**Comment by rightfold** [comment]
@mlawren has any progress been made on this? If not then I don't mind picking it up.

**Comment by mlawren** [comment]
Wow, hadn't remembered that I opened this, so no, no progress has been
made on this ticket. Please go right ahead.

One requirement that you might consider if/when you start on this is to
let the non-SQL code run within an existing transaction or use the
sqitch database handle.  In Perl it could be passed in directly, other
scripts could perhaps read and write commands through STDIN/STDOUT via
sqitch rather than connecting directly themselves.

This behaviour could be important if David merges a recently (Jun 3)
discussed collect/assemble/gather/dump feature, which would bundle a
set of changes together.

**Comment by docteurklein** [comment]
also see https://github.com/theory/sqitch/issues/323#issuecomment-271547337

**Comment by jpotter** [comment]
Hi! We've been using `sqitch` and liking it. However, we're having the same need -- one of our migrations needs to run some non-sql code.

Splitting the migration into 3 sqitch migrations -- one that adds new columns, a second, non-sql one that fills in the new columns (in this case, some python code that does a simple transformation on the data), and a third one that drops the old columns, would be fine.

So a +1 on this particular feature request. I wish I had time to contribute the implementation!

**Comment by theory** [comment]
Yeah, it's a good idea, but there hasn't been much demand. Depending on your database engine, you might well be able to do it in a procedural language in a regular SQL deployment script. For example on Postgres, you could have an earlier change install [PL/Python](https://www.postgresql.org/docs/current/plpython.html) and then, in your script:

```
DO LANGAUGE plpython $$
   # your Python here
$$;
```

But other database servers provide other procedural languages that you might be able to use, too, like [MySQL stored procedures](https://dev.mysql.com/doc/connector-net/en/connector-net-tutorials-stored-procedures.html). I think these database features have tended to erode much demand for non-SQL Sqitch deployment scripts.

**Comment by jpotter** [comment]
Thanks, David! It's a good idea and may work for others -- unfortunately for us, we're running on RDS, so running python procedurally isn't an option. If we do end up patching `sqitch` to support this, I'll let you know; sounds like it's not a priority for others for now. :)
